src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -792,52 +803,97 @@ static int parse_imgpath(const std::string &imgpath)
   return 0;
 }
 
-static int do_list_mapped_devices()
+static int get_mapped_info(int pid)
 {
-  char path[64];
-  int m = 0;
-  int fd[2];
+  int r;
+  std::string path = "/proc/" + stringify(pid) + "/cmdline";
+  std::ifstream ifs;
+  std::string cmdline;
+  std::vector<const char*> args;
 
-  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) == -1) {
-    int r = -errno;
-    cerr << "rbd-nbd: socketpair failed: " << cpp_strerror(-r) << std::endl;
-    return r;
+  ifs.open(path.c_str(), std::ifstream::in);
+  assert (ifs.is_open());
+  ifs >> cmdline;
+
+  for (unsigned int index = 0; index < cmdline.size(); index++) {
+    if (((unsigned char)cmdline[index]) < ' ') {
+      cmdline[index] = ' ';
+      index++;
+    }
   }
 
-  while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
-      break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
+  char *pch = const_cast<char*>(cmdline.c_str());
+  pch = strtok(pch, " ");
+  if (strstr(pch, "rbd-nbd") != NULL) { 
+    pch = strtok(NULL, " ");
+  } else {
+    return EXIT_FAILURE;
   }
 
-  close(fd[0]);
-  close(fd[1]);
+  while (pch != NULL) {
+    args.push_back(pch);
+    pch = strtok(NULL, " ");
+  }
 
-  return 0;
+  r = parse_args(args);
+  return r;
 }
 
-static int rbd_nbd(int argc, const char *argv[])
+static int get_map_pid(const std::string& pid_path)
 {
-  int r;
-  enum {
-    None,
-    Connect,
-    Disconnect,
-    List
-  } cmd = None;
+  int pid = 0;
+  std::ifstream ifs;
+  ifs.open(pid_path.c_str(), std::ifstream::in);
+  if (!ifs.is_open()) {
+    return 0;
+  }
+  ifs >> pid;
+  return pid;
+}
 
-  vector<const char*> args;
+static int do_list_mapped_devices()
+{
+  bool r;
+  bool should_print = false;
+  int index = 0;
+  int pid = 0;
+  TextTable tbl;
 
-  argv_to_vec(argc, argv, args);
-  md_config_t().parse_argv(args);
+  tbl.define_column("pid", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
+
+  while (true) {
+    std::string nbd_path = "/sys/block/nbd" + stringify(index);
+    if(access(nbd_path.c_str(), F_OK) != 0) {
+      break;
+    }
+    std::string pid_path = nbd_path + "/pid";
+    pid = get_map_pid(pid_path);
+
+    if(pid > 0) {
+      r = get_mapped_info(pid);
+      if (r != 0) {
+        index++;
+        continue;
+      } 
+      should_print = true;
+      tbl << stringify(pid) << poolname << imgname << snapname << "/dev/nbd" + stringify(index) << TextTable::endrow;
+    }
 
+    index++;
+  }
+
+  if (should_print) {
+    cout << tbl;
+  }
+  return 0;
+}
+
+static int parse_args(vector<const char*>& args)
```

Comment:
Actually, when building `args` vector from `cmdline` you just need to check the first `arg`, that `basename(arg)` is 'rbd-nbd' and skip pushing it in `args`.